pkg/spoke/managedcluster/healthcheck_controller.go

```diff
@@ -0,0 +1,90 @@
+package managedcluster
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	clientset "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+	clusterv1informer "github.com/open-cluster-management/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	"github.com/open-cluster-management/registration/pkg/helpers"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	discovery "k8s.io/client-go/discovery"
+)
+
+// managedClusterHealthCheckController checks the kube-apiserver health on managed cluster to determine it whether is available.
+type managedClusterHealthCheckController struct {
+	clusterName      string
+	hubClusterClient clientset.Interface
+	hubClusterLister clusterv1listers.ManagedClusterLister
+	discoveryClient  discovery.DiscoveryInterface
+}
+
+// NewManagedClusterHealthCheckController creates a managed cluster health check controller on managed cluster.
+func NewManagedClusterHealthCheckController(
+	clusterName string,
+	hubClusterClient clientset.Interface,
+	hubClusterInformer clusterv1informer.ManagedClusterInformer,
+	discoveryClient discovery.DiscoveryInterface,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterHealthCheckController{
+		clusterName:      clusterName,
+		hubClusterClient: hubClusterClient,
+		hubClusterLister: hubClusterInformer.Lister(),
+		discoveryClient:  discoveryClient,
+	}
+
+	return factory.New().
+		WithInformers(hubClusterInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(1*time.Minute).
+		ToController("ManagedClusterHealthCheckController", recorder)
+}
+
+// sync updates managed cluster available condition by checking kube-apiserver health on managed cluster.
+func (c *managedClusterHealthCheckController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	cluster, err := c.hubClusterLister.Get(c.clusterName)
+	if err != nil {
+		return fmt.Errorf("unable to get managed cluster %q from hub: %w", c.clusterName, err)
+	}
+
+	// current managed cluster is not accepted, do nothing.
```

Comment:
I think if you have credentials, you can try to update this.  I would like to avoid the dependency.

Reply:
Do we want to update this resource directly? Isn't that what we're trying to avoid by using the lease?

Reply:
remove this

Reply:
remove this

Reply:
> Do we want to update this resource directly? Isn't that what we're trying to avoid by using the lease?

This controller to check whether the managed cluster kube-apisever is ready, lease is used to  prove the managed cluster has heartbeat. one case is managed cluster can connect to the hub cluster (has heartbeat), but its kube-apiserver has problem